bindings/node.js/kzg.cxx

```diff
@@ -9,55 +9,90 @@
 #include "c_kzg_4844.h"
 #include "blst.h"
 
-Napi::Value throw_invalid_arguments_count(
-  const unsigned int expected,
-  const unsigned int actual,
-  const Napi::Env env
-) {
-  Napi::RangeError::New(
-    env,
-    "Wrong number of arguments. Expected: "
-    + std::to_string(expected)
-    + ", received " + std::to_string(actual)
-  ).ThrowAsJavaScriptException();
-
-  return env.Null();
-}
+Napi::Value LoadTrustedSetup(const Napi::CallbackInfo &info);
+Napi::Value BlobToKzgCommitment(const Napi::CallbackInfo &info);
+Napi::Value ComputeKzgProof(const Napi::CallbackInfo &info);
+Napi::Value ComputeBlobKzgProof(const Napi::CallbackInfo &info);
+Napi::Value VerifyKzgProof(const Napi::CallbackInfo &info);
+Napi::Value VerifyBlobKzgProof(const Napi::CallbackInfo &info);
+Napi::Value VerifyBlobKzgProofBatch(const Napi::CallbackInfo &info);
+
+class KzgBindings : public Napi::Addon<KzgBindings>
+{
+public:
+  size_t _bytes_per_blob;
+  size_t _bytes_per_commitment;
+  size_t _bytes_per_field_element;
+  size_t _bytes_per_proof;
+  size_t _field_elements_per_blob;
+  std::unique_ptr<KZGSettings> _settings;
+
+  KzgBindings(Napi::Env env, Napi::Object exports) :
+    _bytes_per_blob{BYTES_PER_BLOB},
+    _bytes_per_commitment{BYTES_PER_COMMITMENT},
+    _bytes_per_field_element{BYTES_PER_FIELD_ELEMENT},
+    _bytes_per_proof{BYTES_PER_PROOF},
+    _field_elements_per_blob{FIELD_ELEMENTS_PER_BLOB},
+    _settings{std::make_unique<KZGSettings>()},
+    _is_setup{false} {
+    DefineAddon(exports, {
+      InstanceValue("BYTES_PER_BLOB", Napi::Number::New(env, _bytes_per_blob), napi_enumerable),
+      InstanceValue("BYTES_PER_COMMITMENT", Napi::Number::New(env, _bytes_per_commitment), napi_enumerable),
+      InstanceValue("BYTES_PER_FIELD_ELEMENT", Napi::Number::New(env, _bytes_per_field_element), napi_enumerable),
+      InstanceValue("BYTES_PER_PROOF", Napi::Number::New(env, _bytes_per_proof), napi_enumerable),
+      InstanceValue("FIELD_ELEMENTS_PER_BLOB", Napi::Number::New(env, _field_elements_per_blob), napi_enumerable)
+    });
+    exports["loadTrustedSetup"] = Napi::Function::New(env, LoadTrustedSetup, "setup", this);
+    exports["blobToKzgCommitment"] = Napi::Function::New(env, BlobToKzgCommitment, "blobToKzgCommitment", this);
+    exports["computeKzgProof"] = Napi::Function::New(env, ComputeKzgProof, "computeKzgProof", this);
+    exports["computeBlobKzgProof"] = Napi::Function::New(env, ComputeBlobKzgProof, "computeBlobKzgProof", this);
+    exports["verifyKzgProof"] = Napi::Function::New(env, VerifyKzgProof, "verifyKzgProof", this);
+    exports["verifyBlobKzgProof"] = Napi::Function::New(env, VerifyBlobKzgProof, "verifyBlobKzgProof", this);
+    exports["verifyBlobKzgProofBatch"] = Napi::Function::New(env, VerifyBlobKzgProofBatch, "verifyBlobKzgProofBatch", this);
+  }
+
+  ~KzgBindings() {
```

Comment:
What is this little ~ function? Some sort of destructor? Maybe we can document it as such?

Reply:
Yep, exactly.

Node provides a stable api for plugging in native code to the execution environment called [N-Api](https://nodejs.org/docs/latest-v12.x/api/n-api.html#n_api_napi_set_instance_data). The `Napi` namespace is from a header only c++ implementation library called `node-addon-api` [doclink](https://github.com/nodejs/node-addon-api/blob/main/doc/env.md#setinstancedata).

I did some research and am refactoring to use the underlying C api and a functional paradigm.  It will avoid the c++ destructor and the move/copy constructors and operators that you brought up.  It uses a struct instead of a class and will look and feel much more familiar.

tl/dr;
In reference to the syntax you mentioned.  It will go away but you are 100% correct, that is a c++ destructor.  When a stack allocated instance goes out of scope, or a heap allocated one is `delete`d that function is called (not called by free though). They also can get sneakily called when temporary objects get deleted, if the location is not eligible for copy elison (like return value optimization).

`&&` is a `right-hand` reference (temporary, unnamed variable like the result of an expression between `()`) and the two signatures with it are the move constructor and move-assignment operator-overload.

The other two signatures with `const &` are similar but for copy construction and copy assignment.

By marking all 4 as `delete` it makes the object immovable and non-copiable.

Reply:
@asn-d6 and @jtraglia the first few refactor steps went smoothly and will request a re-review once all the test are green again.

Reply:
ACK. So I guess we are waiting for the refactoring now. No reason to further review the current code.

Reply:
Updated.